components/gatekeeper/auth/AuthServer.go

```diff
@@ -0,0 +1,187 @@
+	}
+	return server
+}
+
+// Default auth check service
+func (s *authServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	log.Infof("Path check, url: %v, path: %v", r.URL, r.URL.Path)
+	// login page open to everyone; all other path requires auth with Password or cookie
+	if strings.HasPrefix(r.URL.Path, "/" + LoginPagePath) || s.authCookie(r) == true {
+		// Handle request from login page
+		if r.Header.Get(LoginPageHeader) != "" {
+			w.WriteHeader(http.StatusResetContent)
+			w.Write([]byte(http.StatusText(http.StatusResetContent)))
+			return
+		}
+		// Allow browser request
+		log.Infof("Allow browser request")
+		w.WriteHeader(http.StatusOK)
+		w.Write([]byte(http.StatusText(http.StatusOK)))
+		return
+	}
+
+	if s.authpwd(r) == true {
+		log.Infof("P/W passed")
+		// Handle request from login page
+		if r.Header.Get(LoginPageHeader) != "" {
+			s.setCookieAndReset(w, r)
+			return
+		}
+		// Allow requst from API call
+		w.WriteHeader(http.StatusOK)
+		w.Write([]byte(http.StatusText(http.StatusOK)))
+		return
+	}
+	log.Infof("Unauthorized, redirect to %v", "https://" + path.Join(r.Host, LoginPagePath))
+
+	// redirect to login page
+	s.redirectToLogin(w, r)
+}
+
+// auth with basic pw
+func (s *authServer) authpwd(r *http.Request) bool {
+	auth := r.Header.Get("Authorization")
+	if !strings.HasPrefix(strings.ToLower(auth), "basic ") {
+		return false
+	}
+
+	upBytes, err := base64.StdEncoding.DecodeString(auth[len("basic "):])
+	if err != nil {
+		return false
+	}
+
+	namepw := strings.Split(string(upBytes), ":")
+
+	if len(namepw) != 2 {
+		return false
+	}
+	err = bcrypt.CompareHashAndPassword([]byte(s.pwhash), []byte(namepw[1]))
+	if namepw[0] == s.username && err == nil {
+
+		return true
+	}
+	return false
+}
+
+// auth with cookie
+func (s *authServer) authCookie(r *http.Request) bool {
+	if cookie, err := r.Cookie(CookieName); err == nil {
+		if val, ok := s.cookies[cookie.Value]; ok {
+			if time.Now().Before(val) {
+				log.Info("cookie auth: passed! %v", cookie.Value)
+				return true
+			}
+			log.Info("cookie auth: cookie value expired!")
+			return false
+		}
+		log.Info("cookie auth: cookie value not found! %v", cookie.Value)
+		return false
+	}
+	log.Info("cookie auth: cookie does't exist in request!")
+	return false
+}
+
+// redirect to login page when unauthorized
+func (s *authServer) redirectToLogin(w http.ResponseWriter, r *http.Request) {
+	http.Redirect(w, r, "https://" + path.Join(r.Host, LoginPagePath), http.StatusTemporaryRedirect)
+}
+
+func generateCookieValue() string {
+	chars := "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+	b := make([]byte, 20)
+	for i := range b {
+		b[i] = chars[rand.Intn(len(chars))]
+	}
+	return string(b)
+}
+
+func (s *authServer) addNewCookieValue(cookieVal string) {
+	s.serverMux.Lock()
+	defer s.serverMux.Unlock()
+	// Cookie expire after 12 hours
+	log.Info("cookie set: set new cookie value!")
+	s.cookies[cookieVal] = time.Now().Add(12 * time.Hour)
```

Comment:
So the cookie is stored in memory, so if the pod is restarted the cookie will no longer be valid. Will we automatically redirect user to log back in?

Reply:
Might be worth a P2 issue to fix in a follow up.